tgui/packages/tgui/interfaces/NtosNotepad.tsx

```diff
@@ -1,39 +1,682 @@
 
-type Data = {
+const DEFAULT_DOCUMENT_NAME = 'Untitled';
+
+interface PartiallyUnderlinedProps {
+  str: string;
+  indexStart: number;
+}
+
+const PartiallyUnderlined = (props: PartiallyUnderlinedProps) => {
+  const { str, indexStart } = props;
+  const start = str.substring(0, indexStart);
+  const underlined = str.substring(indexStart, indexStart + 1);
+  const end = indexStart < str.length - 1 ? str.substring(indexStart + 1) : '';
+  return (
+    <>
+      {start}
+      <span style={{ 'text-decoration': 'underline' }}>{underlined}</span>
+      {end}
+    </>
+  );
+};
+
+enum Dialogs {
+  NONE = 0,
+  UNSAVED_CHANGES = 1,
+  SAVE_AS = 2,
+  OPEN = 3,
+  ABOUT = 4,
+}
+
+type MenuBarProps = {
+  openDocument: () => void;
+  save: () => void;
+  saveAsDialog: () => void;
+  exit: () => void;
+  newNote: () => void;
+  cutSelected: () => void;
+  copySelected: () => void;
+  pasteClipboard: () => void;
+  deleteSelected: () => void;
+  showStatusBar: boolean;
+  setShowStatusBar: (boolean) => void;
+  wordWrap: boolean;
+  setWordWrap: (boolean) => void;
+  aboutNotepadDialog: () => void;
+};
+
+// eslint-disable-function react/jsx-key
+const MenuBar = (props: MenuBarProps, context) => {
+  const {
+    openDocument,
+    save,
+    saveAsDialog,
+    exit,
+    newNote,
+    cutSelected,
+    copySelected,
+    pasteClipboard,
+    deleteSelected,
+    setShowStatusBar,
+    showStatusBar,
+    wordWrap,
+    setWordWrap,
+    aboutNotepadDialog,
+  } = props;
+  const [openOnHover, setOpenOnHover] = useLocalState(
+    context,
+    'openOnHover',
+    false
+  );
+  const [openMenuBar, setOpenMenuBar] = useLocalState<string | null>(
+    context,
+    'openMenuBar',
+    null
+  );
+  const onMenuItemClick = (value) => {
+    setOpenOnHover(false);
+    setOpenMenuBar(null);
+    switch (value) {
+      case 'open':
+        openDocument();
+        break;
+      case 'save':
+        save();
+        break;
+      case 'saveAs':
+        saveAsDialog();
+        break;
+      case 'exit':
+        exit();
+        break;
+      case 'new':
+        newNote();
+        break;
+      case 'cut':
+        cutSelected();
+        break;
+      case 'copy':
+        copySelected();
+        break;
+      case 'paste':
+        pasteClipboard();
+        break;
+      case 'delete':
+        deleteSelected();
+        break;
+      case 'statusBar':
+        setShowStatusBar(!showStatusBar);
+        break;
+      case 'wordWrap':
+        setWordWrap(!wordWrap);
+        break;
+      case 'aboutNotepad':
+        aboutNotepadDialog();
+        break;
+    }
+  };
+  // Adds the key using the value
+  const getMenuItemProps = (value: string, displayText: string) => {
+    return {
+      key: value,
+      value,
+      displayText,
+      onClick: onMenuItemClick,
+    };
+  };
+  /* eslint-disable react/jsx-key */
+  const fileOptions = [
+    <MenuBarDropdown.MenuItem {...getMenuItemProps('new', 'New')} />,
```

Comment:
It's hard to say why this is very wrong, but there is a much simpler way. As I said earlier, it's unnecessary to use arrays. It would also read much better if you simply specified onOpen handler on each of the elements explicitly.

Reply:
It's because MenuBar was originally a Dropdown component. Once I hit the limit with what I could do with Dropdown I had to make a copy of it. The Dropdown component takes an array of "options" so I kind of left as-is but I agree this is ugly.

I'm thinking it would be more intuitive to make MenuBarDropdown take children (the menu items in this case)

Reply:
Yes, would be a big inprovement.